sparse/src/KokkosSparse_SortCrs.hpp

```diff
@@ -338,71 +56,144 @@ void sort_crs_matrix(const execution_space& exec, const rowmap_t& rowmap, const
+        });
   } else {
-    // Try to get teamsize to be largest power of 2 not greater than avg entries
-    // per row
-    // TODO (probably important for performnce): add thread-level sort also, and
-    // use that for small avg degree. But this works for now.
-    lno_t idealTeamSize = 1;
-    lno_t avgDeg        = (entries.extent(0) + numRows - 1) / numRows;
-    while (idealTeamSize < avgDeg / 2) {
-      idealTeamSize *= 2;
+    // On GPUs:
+    //   If the matrix is highly imbalanced, or has long rows AND the dimensions
+    //   are not too large to do one large bulk sort, do that. Otherwise, sort
+    //   using one Kokkos thread per row.
+    Ordinal avgDeg   = (entries.extent(0) + numRows - 1) / numRows;
+    Ordinal maxDeg   = KokkosSparse::Impl::graph_max_degree(exec, rowmap);
+    bool useBulkSort = false;
+    if (KokkosSparse::Impl::useBulkSortHeuristic(avgDeg, maxDeg)) {
+      // Calculate the true number of columns if user didn't pass it in
+      if (numCols == Kokkos::ArithTraits<Ordinal>::max()) {
+        KokkosKernels::Impl::kk_view_reduce_max(exec, entries.extent(0), entries, numCols);
+        numCols++;
+      }
+      uint64_t maxBulkKey = (uint64_t)numRows * (uint64_t)numCols;
+      useBulkSort         = maxBulkKey / numRows == (uint64_t)numCols;
+    }
+    if (useBulkSort) {
+      auto permutation = KokkosSparse::Impl::computeEntryPermutation(exec, rowmap, entries, numCols);
+      // Permutations cannot be done in-place
+      Kokkos::View<typename values_t::value_type*, execution_space> origValues(
+          Kokkos::view_alloc(Kokkos::WithoutInitializing, "origValues"), values.extent(0));
+      Kokkos::View<typename entries_t::value_type*, execution_space> origEntries(
+          Kokkos::view_alloc(Kokkos::WithoutInitializing, "origEntries"), entries.extent(0));
+      Kokkos::deep_copy(exec, origValues, values);
+      Kokkos::deep_copy(exec, origEntries, entries);
+      KokkosSparse::Impl::applyPermutation(exec, permutation, origEntries, entries);
+      KokkosSparse::Impl::applyPermutation(exec, permutation, origValues, values);
+    } else {
+      using TeamPol = Kokkos::TeamPolicy<execution_space>;
+      // Can't use bulk sort approach as matrix dimensions are too large.
+      // Fall back to parallel thread-level sort within each row.
+      Ordinal vectorLength = 1;
+      while (vectorLength < avgDeg / 2) {
+        vectorLength *= 2;
+      }
+      if (vectorLength > TeamPol ::vector_length_max()) vectorLength = TeamPol ::vector_length_max();
+      Impl::MatrixSortThreadFunctor<TeamPol, Ordinal, rowmap_t, entries_t, values_t> funct(numRows, rowmap, entries,
+                                                                                           values);
+      Ordinal teamSize = TeamPol(exec, 1, 1, vectorLength).team_size_recommended(funct, Kokkos::ParallelForTag());
+      Kokkos::parallel_for("sort_crs_matrix",
```

Comment:
Same as above, I think it will be useful for debugging and performance to mention that this runs on CPU vs GPU and with a specific algorithm: bulk vs thread?